File: backend/app/schemas/risk_board.py
"""Response shapes for the Risk analysis board.

Every factor is a ratio of an amount to the sale it sits on (net sale for unflown,
issued sale for the rest), and every threshold that
turns a ratio into a level is sent WITH the response. The page prints them, so a reader
can see why an airline is flagged rather than trusting a colour.
"""
from __future__ import annotations

from datetime import date

from pydantic import BaseModel


class RiskFactorDef(BaseModel):
    key: str
    label: str
    explain: str
    medium: float     # ratio in % at which the factor turns medium
    high: float       # ... and high


class RiskTotals(BaseModel):
    sold: float | None                    # net: issues less refunds
    issued: float | None                  # issues only — the status factors' base
    rows: int
    unflown: float | None
    no_travel_date: float | None
    refunds: float | None                 # signed negative, as the sale carries it
    unmatched_sale: float | None          # a deal ran and none matched — no incentive
    needs_data_sale: float | None         # matched, but cannot be confirmed
    unpriced_sale: float | None           # commission never run
    incentive: float | None               # claimable, counted
    slab_dependent_incentive: float | None
    adm_exposure: float | None            # |debit memos|
    adm_count: int
    acm_credit: float | None
    under_recovery: float | None          # vendor declared less than computed
    declared_mismatch_rows: int
    top_airline: str | None
    top_airline_share_pct: float
    top3_share_pct: float
    hhi: float                            # 0–10,000


class AgingBucket(BaseModel):
    key: str
    label: str
    amount: float
    rows: int


class RiskAirlineRow(BaseModel):
    airline_id: int | None
    airline: str
    sold: float | None
    share_pct: float
    rows: int
    # Each factor as % of this airline's sale.
    factors: dict[str, float]
    levels: dict[str, str]                # key -> low | medium | high
    level: str                            # worst of the above
    score: int                            # 2 per high, 1 per medium — the sort key
    unflown: float | None
    incentive: float | None
    slab_dependent_incentive: float | None
    adm_exposure: float | None


class RiskSummaryResponse(BaseModel):
    scope: str
    currency: str | None
    as_of: date
    date_from: str | None
    date_to: str | None
    factors: list[RiskFactorDef]
    totals: RiskTotals
    overall_levels: dict[str, str]
    overall_factors: dict[str, float]
    concentration_level: str
    unflown_aging: list[AgingBucket]
    by_airline: list[RiskAirlineRow]
    level_counts: dict[str, int]
    other_currencies: dict[str, int]

File: frontend/src/lib/riskBoard.ts
// Client for the Risk analysis board. Mirrors backend/app/schemas/risk_board.py.
import api from "@/lib/api";
import { LIST_PARAMS, type BoardQuery } from "@/lib/salesFlown";

export type RiskLevel = "low" | "medium" | "high";

export interface RiskFactorDef {
  key: string;
  label: string;
  explain: string;
  /** % at which the factor turns medium, and high. Printed on the page. */
  medium: number;
  high: number;
}

export interface RiskTotals {
  sold: number | null;
  issued: number | null;
  rows: number;
  unflown: number | null;
  no_travel_date: number | null;
  refunds: number | null;
  unmatched_sale: number | null;
  needs_data_sale: number | null;
  unpriced_sale: number | null;
  incentive: number | null;
  slab_dependent_incentive: number | null;
  adm_exposure: number | null;
  adm_count: number;
  acm_credit: number | null;
  under_recovery: number | null;
  declared_mismatch_rows: number;
  top_airline: string | null;
  top_airline_share_pct: number;
  top3_share_pct: number;
  hhi: number;
}

export interface AgingBucket {
  key: string;
  label: string;
  amount: number;
  rows: number;
}

export interface RiskAirlineRow {
  airline_id: number | null;
  airline: string;
  sold: number | null;
  share_pct: number;
  rows: number;
  factors: Record<string, number>;
  levels: Record<string, RiskLevel>;
  level: RiskLevel;
  score: number;
  unflown: number | null;
  incentive: number | null;
  slab_dependent_incentive: number | null;
  adm_exposure: number | null;
}

export interface RiskSummary {
  scope: string;
  currency: string | null;
  as_of: string;
  date_from: string | null;
  date_to: string | null;
  factors: RiskFactorDef[];
  totals: RiskTotals;
  overall_levels: Record<string, RiskLevel>;
  overall_factors: Record<string, number>;
  concentration_level: RiskLevel;
  unflown_aging: AgingBucket[];
  by_airline: RiskAirlineRow[];
  level_counts: Record<RiskLevel, number>;
  other_currencies: Record<string, number>;
}

export async function fetchRiskSummary(q: BoardQuery): Promise<RiskSummary> {
  const { data } = await api.get<RiskSummary>("/dashboard/risk/summary", {
    params: q,
    ...LIST_PARAMS,
  });
  return data;
}

/** Status tokens: reserved for state, always shipped with a text label. */
export const LEVEL_META: Record<RiskLevel, { label: string; chip: string; cell: string }> = {
  high: {
    label: "High",
    chip: "bg-red-50 text-red-700 ring-1 ring-red-200",
    cell: "bg-red-50 text-red-800",
  },
  medium: {
    label: "Medium",
    chip: "bg-amber-50 text-amber-800 ring-1 ring-amber-200",
    cell: "bg-amber-50 text-amber-900",
  },
  low: {
    label: "Low",
    chip: "bg-emerald-50 text-emerald-700 ring-1 ring-emerald-200",
    cell: "text-gray-700",
  },
};
